velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp

```diff
@@ -0,0 +1,310 @@
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    extractValues(groups, numGroups, result);
+  }
+
+ private:
+  const int64_t kDefaultExpectedNumItems = 1'000'000;
+  const int64_t kMaxNumItems = 4'000'000;
+  // Spark kMaxNumBits is 67108864, but velox has memory limit sizeClassSizes
+  // 256, so decrease it to not over memory limit.
+  const int64_t kMaxNumBits = 700 * 1024;
+
+  void decodeArguments(
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args) {
+    VELOX_USER_CHECK(args.size() > 0);
+    decodedRaw_.decode(*args[0], rows);
+    if (args.size() > 1) {
+      DecodedVector decodedEstimatedNumItems(*args[1], rows);
+      setConstantArgument(
+          "originalEstimatedNumItems",
+          originalEstimatedNumItems_,
+          decodedEstimatedNumItems);
+      if (args.size() > 2) {
+        VELOX_CHECK_EQ(args.size(), 3);
+        DecodedVector decodedNumBits(*args[2], rows);
+        setConstantArgument(
+            "originalNumBits", originalNumBits_, decodedNumBits);
+      } else {
+        originalNumBits_ = originalEstimatedNumItems_ * 8;
+      }
+    } else {
+      originalEstimatedNumItems_ = kDefaultExpectedNumItems;
+      originalNumBits_ = originalEstimatedNumItems_ * 8;
+    }
+
+    estimatedNumItems_ = std::min(originalEstimatedNumItems_, kMaxNumItems);
+    numBits_ = std::min(originalNumBits_, kMaxNumBits);
+    capacity_ = numBits_ / 16;
+  }
+
+  int32_t getPreAllocatedBufferSize(char** groups, int32_t numGroups) const {
+    int32_t preAllocateSize = 0;
+    for (vector_size_t i = 0; i < numGroups; ++i) {
+      auto group = groups[i];
+      auto accumulator = value<BloomFilterAccumulator>(group);
+      if (UNLIKELY(!accumulator->bloomFilter.isSet())) {
+        continue;
+      }
+
+      auto size = accumulator->serializedSize();
+      StringView serialized;
+      if (StringView::isInline(size)) {
+        continue;
+      }
+      preAllocateSize += size;
+    }
+    return preAllocateSize;
+  }
+
+  static void
+  setConstantArgument(const char* name, int64_t& val, int64_t newVal) {
+    VELOX_USER_CHECK_GT(newVal, 0, "{} must be positive", name);
+    if (val == kMissingArgument) {
+      val = newVal;
+    } else {
+      VELOX_USER_CHECK_EQ(
+          newVal, val, "{} argument must be constant for all input rows", name);
+    }
+  }
+
+  static void setConstantArgument(
+      const char* name,
+      int64_t& val,
+      const DecodedVector& vec) {
+    VELOX_CHECK(
+        vec.isConstantMapping(),
+        "{} argument must be constant for all input rows",
+        name);
+    setConstantArgument(name, val, vec.valueAt<int64_t>(0));
+  }
+
+  static constexpr int64_t kMissingArgument = -1;
+  // Reusable instance of DecodedVector for decoding input vectors.
+  DecodedVector decodedRaw_;
+  DecodedVector decodedIntermediate_;
+  int64_t originalEstimatedNumItems_ = kMissingArgument;
```

Comment:
Do we need both originalEstimatedNumItems_ and estimatedNumItems_ member variables? Looks like just one would be sufficient.

Reply:
Yes, because constant `originalEstimatedNumItems_` is the value in Vector, and it will compare with max value to get `estimatedNumItems_ ` which is used in function

Reply:
Are you saying that estimatedNumItems_ can be lower than originalEstimatedNumItems_ if input value is too large?

Reply:
Yes